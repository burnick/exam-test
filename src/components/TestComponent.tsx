import React from 'react';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from 'theme';
import { Provider } from 'react-redux';
import { persistor, store } from 'store';
import { PersistGate } from 'redux-persist/integration/react';
import { UserContextProvider } from 'components/UserContext';

interface ProvidersProps {
  children: React.ReactNode;
}

const TestComponent = ({ children }: ProvidersProps) => {
  return (
    <ThemeProvider theme={defaultTheme}>
      <Provider store={store}>
        <PersistGate persistor={persistor}>
          <UserContextProvider>{children}</UserContextProvider>
        </PersistGate>
      </Provider>
    </ThemeProvider>
  );
};

export default TestComponent;
