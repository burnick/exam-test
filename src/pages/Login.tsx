import React, { useState, useCallback } from 'react';
import styled from 'styled-components';
import InputText from '../components/InputText';
import Button from '../components/Button';
import * as yup from 'yup';
import { useFormik } from 'formik';
import { LoginFormProps, VariantButton } from 'types';
import { findUser } from 'store/slice/users';
import { useAppDispatch } from 'store/hooks';
import useUserContext from 'components/UserContext';
import { loginFormSchema } from 'utils/schema';

const Login = () => {
  const { foundUser } = useUserContext();
  const dispatch = useAppDispatch();
  const [isSubmitClicked, setIsSubmitClicked] = useState(false);

  const handleSubmit = useCallback(
    (values: LoginFormProps) => {
      dispatch(findUser(values));
      setIsSubmitClicked(true);
    },
    [dispatch],
  );

  const formik = useFormik({
    initialValues: {
      branchId: 0,
      userName: '',
      password: '',
    },
    validationSchema: loginFormSchema,
    onSubmit: handleSubmit,
  });

  return (
    <Container>
      <Title>Login</Title>
      <FormContainer noValidate onSubmit={formik.handleSubmit}>
        <InputText
          placeHolder="Branch Id"
          name="branchId"
          onChange={formik.handleChange}
          value={formik.values.branchId > 0 ? formik.values.branchId : ''}
        />
        <InputText
          placeHolder="User name"
          name="userName"
          onChange={formik.handleChange}
          value={formik.values.userName}
        />
        <InputText
          placeHolder="Password"
          type={'password'}
          name="password"
          onChange={formik.handleChange}
          value={formik.values.password}
        />
        <Button
          label={'login'}
          onClick={formik.submitForm}
          disabled={!formik.isValid && formik.dirty}
          variant={VariantButton.PRIMARY}
        />
      </FormContainer>

      <ErrorContainer role="alert" arial-aria-live="assertive">
        {Object.values(formik.errors).map((error, indexKey) => (
          <span key={indexKey}>{`Error: ${error}`}</span>
        ))}
        {!foundUser && isSubmitClicked ? (
          <span>Error: Username or Password is incorrect</span>
        ) : null}
      </ErrorContainer>
    </Container>
  );
};
const Container = styled.div`
  display: flex;
  justify-content: start;
  height: 350px;
  flex-direction: column;
  width: 300px;
  border: 1.5px solid ${(props) => props.theme.palette.common.black};
  padding: 15px;
  box-sizing: border-box;

  @media only screen and (hover: none) and (pointer: coarse) {
    width: 100%;
    height: 50vh;
  }

  @media only screen and (hover: none) and (pointer: coarse) and (orientation: landscape) {
    width: 50%;
    height: 90vh;
  }
`;

const Title = styled.div`
  font-size: 15px;
  font-weight: bold;
  margin-bottom: 10px;
  flex-direction: row;
`;

const FormContainer = styled.form`
  width: 100%;
  display: flex;
  align-items: start;
  justify-content: start;
  flex-direction: column;
  gap: 15px;
  max-width: 1000px;
`;

const ErrorContainer = styled.div`
  display: flex;
  margin-top: 10px;
  align-items: start;
  justify-content: start;
  flex-direction: column;
  background-color: #fcb8b8;
  color: #ff0000;

  span {
    padding: 5px;
    font-weight: bold;
    font-size: 10px;
  }
`;
export default Login;
