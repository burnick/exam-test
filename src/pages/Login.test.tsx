import React from 'react';
import { cleanup, render, fireEvent, act, waitFor, screen } from '@testing-library/react';
import TestComponent from 'components/TestComponent';
import userEvent from '@testing-library/user-event';
import Login from './Login';

afterEach(cleanup);

test('rendersLogin Component', async () => {
  const { getByRole, getByPlaceholderText } = await render(
    <TestComponent>
      <Login />
    </TestComponent>,
  );
  const buttonElements = getByRole('button', {
    name: /login/i,
  });

  const branchIdElement = getByPlaceholderText('Branch Id');
  const userNameElement = getByPlaceholderText('User name');
  const passwordElement = getByPlaceholderText('Password');

  expect(buttonElements).toBeVisible();
  expect(branchIdElement).toBeVisible();
  expect(passwordElement).toBeVisible();
  expect(userNameElement).toBeVisible();
});

test('test Branch Id Input  for Number input', async () => {
  const { getByPlaceholderText } = render(
    <TestComponent>
      <Login />
    </TestComponent>,
  );
  fireEvent.change(getByPlaceholderText('Branch Id'), {
    target: { value: '123456789' },
  });

  await act(async () => {
    const { getByPlaceholderText } = await render(
      <TestComponent>
        <Login />
      </TestComponent>,
    );
    const inputNode = getByPlaceholderText('Branch Id');
    fireEvent.change(inputNode, { target: { value: '2323' } });
    expect((inputNode as HTMLInputElement).value).toBe('2323');
  });
});

test('test Branch Id Input  for Letter input', async () => {
  const { getByPlaceholderText } = render(
    <TestComponent>
      <Login />
    </TestComponent>,
  );
  fireEvent.change(getByPlaceholderText('Branch Id'), {
    target: { value: '123456789' },
  });

  await act(async () => {
    const { getByPlaceholderText } = await render(
      <TestComponent>
        <Login />
      </TestComponent>,
    );
    const inputNode = getByPlaceholderText('Branch Id');
    fireEvent.change(inputNode, { target: { value: 'abcde' } });
    expect((inputNode as HTMLInputElement).value).toBe('');
  });
});

test('test Button submit for empty form', async () => {
  render(
    <TestComponent>
      <Login />
    </TestComponent>,
  );
  const user = userEvent.setup();

  await user.type(screen.getByPlaceholderText('Branch Id'), 'John');
  await user.click(screen.getByRole('button', { name: /login/i }));
  await waitFor(() =>
    expect(screen.getByRole('button', { name: /login/i })).toBeDisabled(),
  );
});
