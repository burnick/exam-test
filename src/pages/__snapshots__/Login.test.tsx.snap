// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component Test should render correctly 1`] = `ShallowWrapper {}`;
